lib/src/Uncertainty/Algorithm/MetaModel/MetaModelAlgorithm.cxx

```diff
@@ -67,47 +68,124 @@ String MetaModelAlgorithm::__repr__() const
   return oss;
 }
 
+struct TestedDistribution
+{
+  Distribution distribution_;
+  Scalar score_;
+  Scalar bic_;
+  Scalar aic_;
+  Scalar aicc_;
+  String criterion_;
+
+  TestedDistribution(const Distribution & distribution,
+                     const Scalar score,
+                     const Scalar bic,
+                     const Scalar aic,
+                     const Scalar aicc,
+                     const String & criterion)
+  : distribution_(distribution)
+  , score_(score)
+  , bic_(bic)
+  , aic_(aic)
+  , aicc_(aicc)
+  , criterion_(criterion)
+  {
+    // Nothing to do
+  }
+
+  bool operator< (const TestedDistribution& other) const
+  {
+    if (criterion_ == "BIC")
+      return bic_ < other.bic_;
+    else if (criterion_ == "AIC")
+      return aic_ < other.aic_;
+    else if (criterion_ == "AICC")
+      return aicc_ < other.aicc_;
+    else
+      return score_ > other.score_;
+  }
+};
+
 Distribution MetaModelAlgorithm::BuildDistribution(const Sample & inputSample)
 {
   // Recover the distribution, taking into account that we look for performance
   // so we avoid to rebuild expensive distributions as much as possible
   const UnsignedInteger inputDimension = inputSample.getDimension();
-  // For the dependence structure, we use the Spearman independence test to decide between an independent and a Normal copula.
-  Bool isIndependent = true;
-  for (UnsignedInteger j = 0; j < inputDimension && isIndependent; ++ j)
-  {
-    const Sample marginalJ = inputSample.getMarginal(j);
-    for (UnsignedInteger i = j + 1; i < inputDimension && isIndependent; ++ i)
-    {
-      TestResult testResult(HypothesisTest::Spearman(inputSample.getMarginal(i), marginalJ));
-      isIndependent = isIndependent && testResult.getBinaryQualityMeasure();
-    }
-  }
   Collection< Distribution > marginals(inputDimension);
-  // The strategy for the marginals is to find the best continuous 1-d parametric model else fallback to a kernel smoothing
-  KernelSmoothing ks;
+  // The strategy for the marginals is to find the best continuous 1-d parametric model else fallback to a non-parametric one
+  DistributionFactory nonParametricModel;
+  const String nonParametric = ResourceMap::GetAsString("MetaModelAlgorithm-NonParametricModel");
+  if (nonParametric == "Histogram")
+    nonParametricModel = DistributionFactory(HistogramFactory());
+  else
+    nonParametricModel = DistributionFactory(KernelSmoothing());
+  //Collection< DistributionFactory > factories = {NormalFactory(), UniformFactory(), GammaFactory(), ExponentialFactory()}
   Collection< DistributionFactory > factories(DistributionFactory::GetContinuousUniVariateFactories());
```

Comment:
GetContinuousUniVariateFactories also return Histogram, so you might want to exclude it